File: linkedlist/141.c
/**
 * 快慢指针的完美场景
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
bool hasCycle(struct ListNode *head) {
    if(head == NULL) return false;
    struct ListNode *fast, *slow;
    fast = head;
    slow = head;
    while(fast && fast->next)
    {
        fast = fast->next->next;
        slow = slow->next;
        
        if(fast == slow) return true;
    }
    return false;
}

File: linkedlist/141.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func hasCycle(head *ListNode) bool {
    if head == nil {
        return false
    }
    fast := head
    slow := head
    for slow != nil && slow.Next != nil {
        fast = fast.Next
        slow = slow.Next.Next
        if fast == slow {
            return true
        }
    }
    return false
}
